### PR TITLE
[JENKINS-28138] Migration to webhooks with ext point

### PR DESCRIPTION
**Pull request by lanwen**
Working to migrate to webhooks instead of services ([JENKINS-28138](https://issues.jenkins-ci.org/browse/JENKINS-28138))

The plan is
- [x] Incapsulate working with hooks in separate class
  This can help to make process testable and simplify classes with other logic (Cleaner, Triggers)
- [x] Migrate to new webhook manager class in cleaner
- [x] Migrate to new manager in push trigger
- [x] Use in `Cleaner` *queue interface* instead of synchronized block. It can simplify add and fetch operations on how to add for cleaning new names
- [x] Clean old service hooks and replace it with webhooks on job reconfig or manual reregister btn
- [x] Add extension point to contribute new events to hook for some jobs
   New events will be merged with old ones from existent hooks 
- ~~[ ] Extend extension point to contribute hook events to  make possible parse events it interested in~~ (will be done in separate PR)
- [x] Write tests

## Next step is #60 

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/57)

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import static com.google.common.base.Predicates.notNull;
+import static com.google.common.base.Predicates.or;
+import static org.jenkinsci.plugins.github.util.FluentIterableWrapper.from;
+import static org.jenkinsci.plugins.github.util.JobInfoHelpers.associatedNames;
+import static org.jenkinsci.plugins.github.util.JobInfoHelpers.withTrigger;
+import static org.jenkinsci.plugins.github.webhook.GHEventsListener.extractEvents;
+import static org.jenkinsci.plugins.github.webhook.GHEventsListener.isApplicableFor;
+
+/**
+ * @author lanwen (Merkushev Kirill)
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: javadoc required

File: src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+ *         Date: 16.06.15
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: date is useless imho, put `@since TODO`

File: src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+ */
+public class WebhookManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GitHubPushTrigger.class);
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: copy-pasted class for logging?

Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: woops

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: Still not fixed?

Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: it strange, but it fixed in cba5353

File: src/main/java/com/cloudbees/jenkins/Cleaner.java [hunk 4]
```diff
@@ -29,70 +27,52 @@
...
-        return TimeUnit2.MINUTES.toMillis(3);
+        return TimeUnit.MINUTES.toMillis(3);
     }
 
+    /**
+     * Each run this work fetches alive repo names (which has trigger for it)
+     * then if names queue is not empty (any job was reconfigured with GH trigger change),
+     * next name passed to {@link WebhookManager} with list of active names to check and unregister old hooks
+     *
+     * @throws Exception
```
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: useless without description

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java [hunk 5]
```diff
@@ -162,6 +164,12 @@ public void doIndex(StaplerRequest req, StaplerResponse rsp) {
         }
 
         String eventType = req.getHeader("X-GitHub-Event");
+        
+        if("ping".equals(eventType)) {
```
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: space

Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: where?

File: src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        final List<GHEvent> events = from(GHEventsListener.all())
+                .filter(isApplicableFor(project))
+                .transformAndConcat(extractEvents()).toList();
+
+        return new Runnable() {
+            public void run() {
+                LOGGER.info("GitHub webhooks activated for job {} with {} (events: {})",
+                        project.getFullName(), names, events);
+
+                from(names)
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/WebhookManager.java]: ugh..

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import hudson.model.AbstractProject;
+import hudson.triggers.Trigger;
+
+import java.util.Collection;
+
+/**
+ * Utility class which holds converters or predicates (matchers) to filter or convert job lists
+ *
+ * @author lanwen (Merkushev Kirill)
+ * @since 1.11.4
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: this should be 1.12.0 then

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: better place TODO

**Comment by KostyaSha** [comment]
:+1:

File: src/main/java/com/cloudbees/jenkins/Cleaner.java [hunk 8]
```diff
@@ -29,70 +27,50 @@
  */
 @Extension
 public class Cleaner extends PeriodicWork {
-    private final Set<GitHubRepositoryName> couldHaveBeenRemoved = new HashSet<GitHubRepositoryName>();
+    /**
+     * Queue contains repo names prepared to cleanup.
+     * After configure method on job, trigger calls {@link #onStop(AbstractProject)}
+     * which converts to repo names with help of contributors.
+     *
```
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: <p/> ?

File: src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import javax.annotation.CheckReturnValue;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Mostly copypaste from guava's FluentIterable
+ */
+public abstract class FluentIterableWrapper<E> implements Iterable<E> {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java]: Seems this class can be removed in future, mark with no external usage annotation?

File: src/main/java/com/cloudbees/jenkins/Cleaner.java [hunk 10]
```diff
@@ -29,70 +27,50 @@
...
+     * next name passed to {@link WebhookManager} with list of active names to check and unregister old hooks
+     */
     @Override
     protected void doRun() throws Exception {
-        List<GitHubRepositoryName> names;
-        synchronized (this) {// atomically obtain what we need to check
-            names = new ArrayList<GitHubRepositoryName>(couldHaveBeenRemoved);
-            couldHaveBeenRemoved.clear();
-        }
+        URL url = Trigger.all().get(GitHubPushTrigger.DescriptorImpl.class).getHookUrl();
```
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: Skip run if queue not empty to allow end previous run?

File: src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Returns an {@code ImmutableSet} containing all of the elements from this fluent iterable with
+     * duplicates removed.
+     *
+     */
+    public final ImmutableSet<E> toSet() {
+        return ImmutableSet.copyOf(iterable);
+    }
+
+
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java]: new line

File: src/main/java/org/jenkinsci/plugins/github/webhook/GHEventsListener.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+ * Extension point to contribute events plugin interested in.
+ * This point should return true in {@link #isApplicable(AbstractProject)}
+ * only if it can parse hooks with events contributed in {@link #events()}
+ *
+ * Each time this plugin wants to get events list from contributors it asks for applicable status
+ *
+ * @author lanwen (Merkushev Kirill)
+ * @since TODO
+ */
+public abstract class GHEventsListener implements ExtensionPoint {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/github/webhook/GHEventsListener.java]: This extension point slightly differs from common Jenkins ones. Listener is supposed to get events somehow. I think it's "Subscriber", not "Listener"

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        };
+    }
+
+    /**
+     * Can be useful to ignore disabled jobs on reregistering hooks
+     *
+     * @return predicate with true on apply if job is buildable
+     */
+    public static Predicate<AbstractProject> isBuildable() {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: Probably it's better to start supporting <code>Job</code>s instead of AbstractProject

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java [hunk 14]
```diff
@@ -162,6 +164,12 @@ public void doIndex(StaplerRequest req, StaplerResponse rsp) {
         }
 
         String eventType = req.getHeader("X-GitHub-Event");
+
+        if ("ping".equals(eventType)) {
+            LOGGER.info("Got a ping request from GitHub");
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: Should be FINE IMO

File: src/main/java/org/jenkinsci/plugins/github/webhook/GHEventsListener.java (hunk 12 above)
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/GHEventsListener.java]: Imho this should be part of Contributor...

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java (hunk 14 above)
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: Ping request is manual thing and must be available in jenkins log

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java (hunk 13 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: it needs to bump core version to do so, isn't it?

File: src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2008 The Guava Authors
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java]: Usage of Apache code in the MIT plugin requires additional notes on the top level.
Was it possible not to copy the code?

File: src/main/java/com/cloudbees/jenkins/Cleaner.java (hunk 10 above)
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: think this logic should be investigated in separate pr

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java (hunk 13 above)
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: Job has no triggers and this adds overhead with type checking

File: src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java (hunk 15 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java]: only if upgrade and shade guava 18

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java]: This is a temporary solution that will be removed when jenkins will have newer dependencies or java8.

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java (hunk 13 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: No, <code>Job</code> exists in Jenkins for a long-long time. So it's better to use it in util methods wherever it's possible in order to simplify further integrations when and if they happen.

**Comment by oleg-nenashev** [comment]
In general the code looks good to me (special kudos for extra tests). If the plugin works well (a manual testing is required), I would vote for such implementation

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java (hunk 14 above)
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: Hmm... I'f there's no automatic call from GitHub, I'm OK with fine()

File: src/main/java/com/cloudbees/jenkins/Cleaner.java (hunk 8 above)
Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: think nobody will be read this in generated javadoc format. Only here in code. So don't like `<p/>`

File: src/main/java/com/cloudbees/jenkins/Cleaner.java (hunk 10 above)
Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: Just one check 
```
if (!queue.isEmpty()) { return; }
```

Comment (lanwen) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: so when first work finds non empty queue it just skip it. And no one will clear it ever :)

Comment (KostyaSha) [inline src/main/java/com/cloudbees/jenkins/Cleaner.java]: probably i mistaken. i thought that it handle only full run of cleaning (not single).

File: src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java (hunk 15 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/github/util/FluentIterableWrapper.java]: OK, agreed. Just reference the code usage in README.md as an acknoledgement

File: pom.xml [hunk 16]
```diff
@@ -34,39 +35,74 @@
...
+          <scope>test</scope>
+      </dependency>
+
+      <dependency>
+          <groupId>org.mockito</groupId>
+          <artifactId>mockito-core</artifactId>
+          <version>1.10.19</version>
+          <scope>test</scope>
+      </dependency>
+
```
Comment (jglick) [inline pom.xml]: Are you actually changing dependencies in this diff hunk? If so, which? Please avoid gratuitous reformatting of files—keep the diff as short as possible. Makes your PR easier to review, and avoids unnecessary merge conflicts, cherry-pick conflicts, `git-blame` noise, etc.

Comment (KostyaSha) [inline pom.xml]: Jessy, i'm "maintaining" last releases. I'm ok with this change. Also there is no any cherry-picks/merge conflicts/etc for this plugin.

git-blame and history will be more noisy when one small PR has 10 commits.

Comment (lanwen) [inline pom.xml]: @jglick, I know about small diffs, thx! But this change alreay discussed with @KostyaSha

**Comment by lanwen** [comment]
Manually tested over real GH

Log:
```
июл 12, 2015 1:37:06 AM INFO org.jenkinsci.plugins.github.webhook.WebhookManager$1 run
GitHub webhooks activated for job test with [GitHubRepository[host=github.com,username=lanwen,repository=test]] (events: [PUSH])
июл 12, 2015 1:37:08 AM FINE org.jenkinsci.plugins.github.webhook.WebhookManager$2 apply
Hook already registered for events [PUSH]
```

or

```
июл 12, 2015 1:41:05 AM FINE org.jenkinsci.plugins.github.webhook.WebhookManager unregisterFor
Check Repository:lanwen:test for redundant hooks...
июл 12, 2015 1:41:05 AM FINE org.jenkinsci.plugins.github.webhook.WebhookManager$3 apply
Deleted hook https://api.github.com/repos/lanwen/test/hooks/5281772 (events: [PUSH])
```

then

```
GitHub webhooks activated for job test with [GitHubRepository[host=github.com,username=lanwen,repository=test]] (events: [PUSH])
июл 12, 2015 1:42:58 AM FINE org.jenkinsci.plugins.github.webhook.WebhookManager$3 apply
Created hook https://api.github.com/repos/lanwen/test/hooks/5281821 (events: [PUSH])
июл 12, 2015 1:42:59 AM FINE org.jenkinsci.plugins.github.webhook.GHEventHeader$PayloadHandler parse
Header X-GitHub-Event -> ping
июл 12, 2015 1:42:59 AM FINE org.jenkinsci.plugins.github.webhook.GHEventHeader$PayloadHandler parse
Unknown event - No enum constant org.kohsuke.github.GHEvent.PING
июл 12, 2015 1:42:59 AM FINEST org.jenkinsci.plugins.github.webhook.GHEventPayload$PayloadHandler parse
Payload {"zen":"Speak like a human.","hook_id":5281821,"hook":{"url":"https://api.github.com/repos/lanwen/test/hooks/5281821","test_url":"https://api.github.com/repos/lanwen/test/hooks/5281821/test","ping_url":"https://api.github.com/repos/lanwen/test/hooks/5281821/pings","id":5281821,"name":"web","active":true,"events":["push"],"config":{"url":"http://3352fb71.ngrok.com/jenkins/github-webhook/"},"last_response":{"code":null,"status":"unused","message":null},"updated_at":"2015-07-11T22:42:57Z","created_at":"2015-07-11T22:42:57Z"},"repository":{"id":38941520,"name":"test","full_name":"lanwen/test","owner":{"login":"lanwen","id":1964214,"avatar_url":"https://avatars.githubusercontent.com/u/1964214?v=3","gravatar_id":"","url":"https://api.github.com/users/lanwen","html_url":"https://github.com/lanwen","followers_url":"https://api.github.com/users/lanwen/followers","following_url":"https://api.github.com/users/lanwen/following{/other_user}","gists_url":"https://api.github.com/users/lanwen/gists{/gist_id}","starred_url":"https://api.github.com/users/lanwen/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/lanwen/subscriptions","organizations_url":"https://api.github.com/users/lanwen/orgs","repos_url":"https://api.github.com/users/lanwen/repos","events_url":"https://api.github.com/users/lanwen/events{/privacy}","received_events_url":"https://api.github.com/users/lanwen/received_events","type":"User","site_admin":false},"private":false,"html_url":"https://github.com/lanwen/test","description":"for test purposes","fork":false,"url":"https://api.github.com/repos/lanwen/test","forks_url":"https://api.github.com/repos/lanwen/test/forks","keys_url":"https://api.github.com/repos/lanwen/test/keys{/key_id}","collaborators_url":"https://api.github.com/repos/lanwen/test/collaborators{/collaborator}","teams_url":"https://api.github.com/repos/lanwen/test/teams","hooks_url":"https://api.github.com/repos/lanwen/test/hooks","issue_events_url":"https://api.github.com/repos/lanwen/test/issues/events{/number}","events_url":"https://api.github.com/repos/lanwen/test/events","assignees_url":"https://api.github.com/repos/lanwen/test/assignees{/user}","branches_url":"https://api.github.com/repos/lanwen/test/branches{/branch}","tags_url":"https://api.github.com/repos/lanwen/test/tags","blobs_url":"https://api.github.com/repos/lanwen/test/git/blobs{/sha}","git_tags_url":"https://api.github.com/repos/lanwen/test/git/tags{/sha}","git_refs_url":"https://api.github.com/repos/lanwen/test/git/refs{/sha}","trees_url":"https://api.github.com/repos/lanwen/test/git/trees{/sha}","statuses_url":"https://api.github.com/repos/lanwen/test/statuses/{sha}","languages_url":"https://api.github.com/repos/lanwen/test/languages","stargazers_url":"https://api.github.com/repos/lanwen/test/stargazers","contributors_url":"https://api.github.com/repos/lanwen/test/contributors","subscribers_url":"https://api.github.com/repos/lanwen/test/subscribers","subscription_url":"https://api.github.com/repos/lanwen/test/subscription","commits_url":"https://api.github.com/repos/lanwen/test/commits{/sha}","git_commits_url":"https://api.github.com/repos/lanwen/test/git/commits{/sha}","comments_url":"https://api.github.com/repos/lanwen/test/comments{/number}","issue_comment_url":"https://api.github.com/repos/lanwen/test/issues/comments{/number}","contents_url":"https://api.github.com/repos/lanwen/test/contents/{+path}","compare_url":"https://api.github.com/repos/lanwen/test/compare/{base}...{head}","merges_url":"https://api.github.com/repos/lanwen/test/merges","archive_url":"https://api.github.com/repos/lanwen/test/{archive_format}{/ref}","downloads_url":"https://api.github.com/repos/lanwen/test/downloads","issues_url":"https://api.github.com/repos/lanwen/test/issues{/number}","pulls_url":"https://api.github.com/repos/lanwen/test/pulls{/number}","milestones_url":"https://api.github.com/repos/lanwen/test/milestones{/number}","notifications_url":"https://api.github.com/repos/lanwen/test/notifications{?since,all,participating}","labels_url":"https://api.github.com/repos/lanwen/test/labels{/name}","releases_url":"https://api.github.com/repos/lanwen/test/releases{/id}","created_at":"2015-07-11T21:47:22Z","updated_at":"2015-07-11T21:47:22Z","pushed_at":"2015-07-11T22:21:07Z","git_url":"git://github.com/lanwen/test.git","ssh_url":"git@github.com:lanwen/test.git","clone_url":"https://github.com/lanwen/test.git","svn_url":"https://github.com/lanwen/test","homepage":null,"size":0,"stargazers_count":0,"watchers_count":0,"language":null,"has_issues":true,"has_downloads":true,"has_wiki":true,"has_pages":false,"forks_count":0,"mirror_url":null,"open_issues_count":0,"forks":0,"open_issues":0,"watchers":0,"default_branch":"master"},"sender":{"login":"lanwen","id":1964214,"avatar_url":"https://avatars.githubusercontent.com/u/1964214?v=3","gravatar_id":"","url":"https://api.github.com/users/lanwen","html_url":"https://github.com/lanwen","followers_url":"https://api.github.com/users/lanwen/followers","following_url":"https://api.github.com/users/lanwen/following{/other_user}","gists_url":"https://api.github.com/users/lanwen/gists{/gist_id}","starred_url":"https://api.github.com/users/lanwen/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/lanwen/subscriptions","organizations_url":"https://api.github.com/users/lanwen/orgs","repos_url":"https://api.github.com/users/lanwen/repos","events_url":"https://api.github.com/users/lanwen/events{/privacy}","received_events_url":"https://api.github.com/users/lanwen/received_events","type":"User","site_admin":false}}
июл 12, 2015 1:42:59 AM INFO org.jenkinsci.plugins.github.webhook.RequirePostWithGHHookPayload$Processor logPingEvent
Got ping event from GH
```

**Comment by amuniz** [comment]
Does this PR require #60 to be merged or vice versa?

**Comment by lanwen** [comment]
@amuniz yes, this feature was splitted for two prs to simplify review.

File: src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+
+    /**
+     * If any of event subscriber interested in hook for job, then return true
+     * By default, push hook subscriber is interested in job with gh-push-trigger
+     * 
+     * @return predicate with true if job alive and should have hook 
+     */
+    public static Predicate<AbstractProject> isAlive() {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: rename to `isApplicable()`?

Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java]: this means that job alive in terms of github hook. So "applicable" word is not clarify meaning
